### DISCUSSION TITLE
ORM mode broken in Pydantic v2

### DISCUSSION THREAD

**Discussion by stinovlas** [pydantic/pydantic#7669] [Question] [answered]
`from_orm` has been deprecated, but not removed in Pydantic v2. However, its behaviour changed significantly. Consider following code:

## Pydantic v1
```python
from pydantic import BaseModel

class A:
    value: int = 10

class B:
    a: A = A()

class C:
    inner: B = B()

class InnerModel(BaseModel):
    value: int

    class Config:
        orm_mode = True

    @classmethod
    def from_orm(cls, obj):
        return cls(value=obj.a.value)

class OuterModel(BaseModel):
    inner: InnerModel

    class Config:
        orm_mode = True

print(OuterModel.from_orm(C()))
```

This runs without errors and outputs `inner=InnerModel(value=10)`.

## Pydantic v2
```python
from pydantic import BaseModel, ConfigDict

class A:
    value: int = 10

class B:
    a: A = A()

class C:
    inner: B = B()

class InnerModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    value: int

    @classmethod
    def from_orm(cls, obj):
        return cls(value=obj.a.value)

class OuterModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    inner: InnerModel

print(OuterModel.from_orm(C()))
```
This fails with following error:

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for OuterModel
inner.value
  Field required [type=missing, input_value=<__main__.B object at 0x7f4621026e90>, input_type=B]
    For further information visit https://errors.pydantic.dev/2.4/v/missing
```

## Result

The `InnerModel.from_orm` is not called at all. I tried to replace `from_orm` with `model_validate`, but it doesn't call inner class' `model_validate` either, so that's a no go.

Is there any way to achieve this behavior in Pydantic v2? Changes in Pydantic v2 have broken `from_orm` and I don't see an obvious replacement.

More generally: we do have quite a deep, nested ORM models. We have much flatter API. We'd like to flatten the ORM objects using `from_orm` or some similar technique, just as we were able to do in Pydantic v1.

**Comment by ImogenBits** [comment]
In general the best way to to model wide validation steps is to use [model validators](https://docs.pydantic.dev/2.4/concepts/validators/#model-validators), but your use case is much easier by just using an [AliasPath](https://docs.pydantic.dev/2.4/concepts/fields/#aliaspath-and-aliaschoices). Your code would then look like this:
```py
from pydantic import AliasPath, BaseModel, ConfigDict, Field

class A:
    value: int = 10

class B:
    a: A = A()

class C:
    inner: B = B()

class InnerModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    value: int = Field(validation_alias=AliasPath("a", "value"))

class OuterModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    inner: InnerModel

print(OuterModel.model_validate(C()))
```

**Reply by stinovlas** [reply]
Thank you! Although it doesn't explain the breakage of `from_orm` between v1 and v2, this actually helped a lot to resolve my usecase! Model validators are allmighty, but they can be quite complicated, `AliasPath` is great (especially when combined with `populate_by_name`).

**Reply by ImogenBits** [reply, marked as answer]
The reason why it doesn't work like it did in V1 is that model validation is now done with the pydantic-core rust package. So instead of calling each submodel's model_validate (or from_orm) recursively in python, only the outermost model's method is called. The control flow then is passed to rust code, which performs validation purely based on the pydantic-core schema of this outermost model. If you want to have hooks back into python code you need to explicitly include them in the pydantic-core schema, either via the __get_pydantic_core_schema__ method, model/field validators, Annotated metadata, or similar pydantic mechanisms. You cannot rely on the python method name resolution to find your custom method since these lookups are never attempted.

Note that this is the intended behaviour and cannot be "fixed", the whole reason why V2 is so much faster and offers other great benefits is that validation logic is performed in rust based on the core schema.

**Reply by stinovlas** [reply]
Thank you! I'm gonna mark this as an answer, because it addresses the original question, but thank you again for pointing out `AliasPath`.

We have quite large ecosystem of libraries and applications that use those libraries and we use pydantic throughout most of it. Therefore, it's not viable to just migrate everything at once. We need to create v1/v2 compatible libraries first, then migrate our applications and in turn migrate our libraries to v2 only, from top to bottom. `pydantic.v1` module has been a great help, but it doesn't solve everything, sadly.

**Comment by samuelcolvin** [comment]
Thanks for the great explanation @ImogenBits.

Sorry you're having trouble @stinovlas, hopefully it's worth it in the end.

**Reply by stinovlas** [reply]
Thank you for creating and maintaining `pydantic`! I'm sure we'll overcome the problems given enough time.
